server/app.py

```diff
@@ -307,6 +311,81 @@ async def ready(redis_conn: Redis = Depends(_get_redis_conn)):
     return {"status": "ok"}
 
 
+class ImprovedPromptRequest(BaseModel):
+    """
+    Request model for improving a prompt.
+    """
+    student_skill: Skill
+    student_model: str
+    teacher_runtime: AsyncRuntime
+    input_variables: List[str]
+    
+    # same code as for ResultHandler in SubmitStreamingRequest
+    @field_validator("student_skill", mode="before")
+    def validate_skill(cls, value: Dict) -> Skill:
+        if "type" not in value:
+            raise HTTPException(
+                status_code=400, detail="Missing type in student_skill"
+            )
+        skill = Skill.create_from_registry(value.pop("type"), **value)
+        return skill
+    
+    # same code as for ResultHandler in SubmitStreamingRequest
+    @field_validator("teacher_runtime", mode="before")
+    def validate_teacher_runtime(cls, value: Dict) -> AsyncRuntime:
+        if "type" not in value:
+            raise HTTPException(
+                status_code=400, detail="Missing type in teacher_runtime"
+            )
+        runtime = AsyncRuntime.create_from_registry(value.pop("type"), **value)
+        return runtime
+
+class ImprovedPromptResponse(BaseModel):
+    
+    output: Optional[PromptImprovementSkillResponseModel] = None
+    
+    prompt_tokens: int = Field(alias="_prompt_tokens")
+    completion_tokens: int = Field(alias="_completion_tokens")
+
+    # these can fail to calculate
+    prompt_cost_usd: Optional[float] = Field(alias="_prompt_cost_usd")
+    completion_cost_usd: Optional[float] = Field(alias="_completion_cost_usd")
+    total_cost_usd: Optional[float] = Field(alias="_total_cost_usd")
+
+@app.post("/improved-prompt", response_model=Response[ImprovedPromptResponse])
+async def improved_prompt(request: ImprovedPromptRequest):
+    """
+    Improve a given prompt using the specified model and variables.
+
+    Args:
+        request (ImprovedPromptRequest): The request model for improving a prompt.
+
+    Returns:
+        Response: Response model for prompt improvement skill
+    """
+    
+    inputs = get_prompt_improvement_inputs(request.student_skill, request.input_variables, request.student_model)
```

Comment:
why not making it internal function of `Agent` ?

Reply:
good point. The info that doesn't fit in the Agent is the `input_variables`, because that's a list of all variables (from the project) that _could_ be used in the prompt regardless of if they currently are, and the Environment in the agent would not be used. But it's close enough that it could go there.

Reply:
done